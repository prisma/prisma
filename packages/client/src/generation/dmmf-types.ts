import { DMMF } from '@prisma/generator-helper'

export { DMMF }

export type BaseDMMF = {
  readonly datamodel: Omit<DMMF.Datamodel, 'indexes'>
}
