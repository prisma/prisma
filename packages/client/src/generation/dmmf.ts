import type { DMMF } from '@prisma/generator-helper'

import type { Dictionary } from './utils/common'
import { keyBy } from './utils/common'

type NamespacedTypeMap<T> = {
  prisma: Record<string, T>
  model: Record<string, T>
}

type FullyQualifiedName = string & { readonly _brand: unique symbol }

export class DMMFHelper implements DMMF.Document {
  private _compositeNames?: Set<string>
  private _inputTypesByName?: Map<FullyQualifiedName, DMMF.InputType>
  private _typeAndModelMap?: Dictionary<DMMF.Model>
  private _mappingsMap?: Dictionary<DMMF.ModelMapping>
  private _outputTypeMap?: NamespacedTypeMap<DMMF.OutputType>
  private _rootFieldMap?: Dictionary<DMMF.SchemaField>

  constructor(public document: DMMF.Document) {}

  private get compositeNames(): Set<string> {
    return (this._compositeNames ??= new Set(this.datamodel.types.map((t) => t.name)))
  }

  private get inputTypesByName(): Map<FullyQualifiedName, DMMF.InputType> {
    return (this._inputTypesByName ??= this.buildInputTypesMap())
  }

  get typeAndModelMap(): Dictionary<DMMF.Model> {
    return (this._typeAndModelMap ??= this.buildTypeModelMap())
  }

  get mappingsMap(): Dictionary<DMMF.ModelMapping> {
    return (this._mappingsMap ??= this.buildMappingsMap())
  }

  get outputTypeMap(): NamespacedTypeMap<DMMF.OutputType> {
    return (this._outputTypeMap ??= this.buildMergedOutputTypeMap())
  }

  get rootFieldMap(): Dictionary<DMMF.SchemaField> {
    return (this._rootFieldMap ??= this.buildRootFieldMap())
  }

  get datamodel(): DMMF.Datamodel {
    return this.document.datamodel
  }

  get mappings(): DMMF.Mappings {
    return this.document.mappings
  }

  get schema(): DMMF.Schema {
    return this.document.schema
  }

  get inputObjectTypes(): DMMF.Schema['inputObjectTypes'] {
    return this.schema.inputObjectTypes
  }

  get outputObjectTypes(): DMMF.Schema['outputObjectTypes'] {
    return this.schema.outputObjectTypes
  }

  isComposite(modelOrTypeName: string) {
    return this.compositeNames.has(modelOrTypeName)
  }

  getOtherOperationNames() {
    return [
      Object.values(this.mappings.otherOperations.write),
      Object.values(this.mappings.otherOperations.read),
    ].flat()
  }

  hasEnumInNamespace(enumName: string, namespace: DMMF.FieldNamespace): boolean {
    return this.schema.enumTypes[namespace]?.find((schemaEnum) => schemaEnum.name === enumName) !== undefined
  }

  resolveInputObjectType(ref: DMMF.InputTypeRef): DMMF.InputType | undefined {
    return this.inputTypesByName.get(fullyQualifiedName(ref.type, ref.namespace))
  }

  resolveOutputObjectType(ref: DMMF.OutputTypeRef): DMMF.OutputType | undefined {
    if (ref.location !== 'outputObjectTypes') {
      return undefined
    }
    return this.outputObjectTypes[ref.namespace ?? 'prisma'].find((outputObject) => outputObject.name === ref.type)
  }

  private buildModelMap(): Dictionary<DMMF.Model> {
    return keyBy(this.datamodel.models, 'name')
  }

  private buildTypeMap(): Dictionary<DMMF.Model> {
    return keyBy(this.datamodel.types, 'name')
  }

  private buildTypeModelMap(): Dictionary<DMMF.Model> {
    return { ...this.buildTypeMap(), ...this.buildModelMap() }
  }

  private buildMappingsMap(): Dictionary<DMMF.ModelMapping> {
    return keyBy(this.mappings.modelOperations, 'model')
  }

  private buildMergedOutputTypeMap(): NamespacedTypeMap<DMMF.OutputType> {
    return {
      model: keyBy(this.schema.outputObjectTypes.model, 'name'),
      prisma: keyBy(this.schema.outputObjectTypes.prisma, 'name'),
    }
  }

  private buildRootFieldMap(): Dictionary<DMMF.SchemaField> {
    return {
      ...keyBy(this.outputTypeMap.prisma.Query.fields, 'name'),
      ...keyBy(this.outputTypeMap.prisma.Mutation.fields, 'name'),
    }
  }

  private buildInputTypesMap() {
    const result = new Map<FullyQualifiedName, DMMF.InputType>()
    for (const type of this.inputObjectTypes.prisma) {
      result.set(fullyQualifiedName(type.name, 'prisma'), type)
    }

    if (!this.inputObjectTypes.model) {
      return result
    }

    for (const type of this.inputObjectTypes.model) {
      result.set(fullyQualifiedName(type.name, 'model'), type)
    }
    return result
  }
}

function fullyQualifiedName(typeName: string, namespace?: string) {
  if (namespace) {
    return `${namespace}.${typeName}` as FullyQualifiedName
  }
  return typeName as FullyQualifiedName
}
