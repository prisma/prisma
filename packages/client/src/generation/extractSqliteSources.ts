import { absolutizeRelativePath } from '../utils/resolveDatasources'

export interface DatasourceOverwrite {
  name: string
  url?: string
  env?: string
}

// only extract sqlite sources that don't use env vars
export function extractSqliteSources(
  datamodel: string,
  cwd: string,
  outputDir: string,
  absolutePaths?: boolean,
): DatasourceOverwrite[] {
  const overrides: DatasourceOverwrite[] = []
  const lines = datamodel.split('\n').filter((l) => !l.trim().startsWith('//'))
  const lineRegex = /\s*url\s*=\s*("(file|sqlite):([^\/].*)"|env\("(\w+)"\))/
  const startRegex = /\s*datasource\s*(\w+)\s*{/

  lines.forEach((line, index) => {
    const match = lineRegex.exec(line)
    if (match) {
      // search for open tag
      let startLine
      let searchIndex = index - 1
      while (!startLine && searchIndex >= 0) {
        const currentLine = lines[searchIndex]
        const commentIndex = currentLine.indexOf('//')
        const curlyIndex = currentLine.indexOf('{')
        if (curlyIndex > -1) {
          if (commentIndex === -1) {
            startLine = currentLine
          }
          if (commentIndex > curlyIndex) {
            startLine = currentLine
          }
        }

        searchIndex--
      }

      if (!startLine) {
        throw new Error(`Could not parse datamodel, invalid datasource block without opening \`{\``)
      }

      const startMatch = startRegex.exec(startLine)
      if (startMatch) {
        if (match[4]) {
          overrides.push({
            name: startMatch[1],
            env: match[4],
          })
        } else {
          overrides.push({
            name: startMatch[1],
            url: absolutizeRelativePath(match[3], cwd, outputDir, absolutePaths),
          })
        }
      } else {
        throw new Error(`Could not parse datamodel, line ${searchIndex + 1}: \`${startLine}\` is not parseable`)
      }
    }
  })
  return overrides
}
