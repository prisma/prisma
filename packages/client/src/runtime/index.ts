import * as lzString from 'lz-string'

import * as Extensions from './core/extensions'
import * as Types from './core/types'

export { PrismaClientExtensionError } from './core/extensions/wrapExtensionCallback'
export {
  type Metric,
  type MetricHistogram,
  type MetricHistogramBucket,
  type Metrics,
  MetricsClient,
} from './core/metrics/MetricsClient'
export type { FieldRef } from './core/model/FieldRef'
export { DMMFHelper as DMMFClass } from './dmmf'
export { type BaseDMMF, DMMF } from './dmmf-types'
export type { PrismaClientOptions } from './getPrismaClient'
export { getPrismaClient } from './getPrismaClient'
export { objectEnumValues } from './object-enums'
export { makeDocument, PrismaClientValidationError, transformDocument, unpack } from './query'
export { makeStrictEnum } from './strictEnum'
export type { DecimalJsLike } from './utils/decimalJsLike'
export { findSync } from './utils/find'
export { NotFoundError } from './utils/rejectOnNotFound'
export { warnEnvConflicts } from './warnEnvConflicts'
export { Debug } from '@prisma/debug'
export {
  Engine,
  PrismaClientInitializationError,
  PrismaClientKnownRequestError,
  PrismaClientRustPanicError,
  PrismaClientUnknownRequestError,
} from '@prisma/engine-core'
export { default as Decimal } from 'decimal.js'
export type { RawValue, Value } from 'sql-template-tag'
export { empty, join, raw, Sql, default as sqltag } from 'sql-template-tag'
// ! export bundling fails for this dep, we work around it
const decompressFromBase64 = lzString.decompressFromBase64
export { decompressFromBase64 }

export { Types }
export { Extensions }
