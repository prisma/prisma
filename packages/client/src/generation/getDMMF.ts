import type { DMMF } from '@prisma/generator-helper'
import type { GetDMMFOptions } from '@prisma/internals'
import { getDMMF as getRawDMMF } from '@prisma/internals'

import type { DMMF as PrismaClientDMMF } from '../runtime/dmmf-types'
import { externalToInternalDmmf } from '../runtime/externalToInternalDmmf'

export function getPrismaClientDMMF(dmmf: DMMF.Document): PrismaClientDMMF.Document {
  return externalToInternalDmmf(dmmf)
}

// Mostly used for tests
export async function getDMMF(options: GetDMMFOptions): Promise<PrismaClientDMMF.Document> {
  const dmmf = await getRawDMMF(options)
  return getPrismaClientDMMF(dmmf)
}
