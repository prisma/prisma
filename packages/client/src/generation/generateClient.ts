import { BinaryType, overwriteFile } from '@prisma/fetch-engine'
import type { BinaryPaths, DataSource, DMMF, GeneratorConfig } from '@prisma/generator-helper'
import type { Platform } from '@prisma/internals'
import { ClientEngineType, getClientEngineType, getEngineVersion } from '@prisma/internals'
import copy from '@timsuchanek/copy'
import chalk from 'chalk'
import fs from 'fs'
import { ensureDir } from 'fs-extra'
import path from 'path'
import pkgUp from 'pkg-up'
import type { O } from 'ts-toolbelt'
import { promisify } from 'util'

import { name as clientPackageName } from '../../package.json'
import type { DMMF as PrismaClientDMMF } from '../runtime/dmmf-types'
import type { Dictionary } from '../runtime/utils/common'
import { getPrismaClientDMMF } from './getDMMF'
import { BrowserJS, JS, TS, TSClient } from './TSClient'

const remove = promisify(fs.unlink)
const writeFile = promisify(fs.writeFile)
const exists = promisify(fs.exists)
const copyFile = promisify(fs.copyFile)
const stat = promisify(fs.stat)

const GENERATED_PACKAGE_NAME = '.prisma/client'

type OutputDeclaration = {
  content: string
  lineNumber: number
}

export class DenylistError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'DenylistError'
    this.stack = undefined
  }
}

export interface GenerateClientOptions {
  projectRoot?: string
  datamodel: string
  schemaPath: string
  transpile?: boolean
  runtimeDirs?: { node: string; edge: string }
  outputDir: string
  generator?: GeneratorConfig
  dmmf: DMMF.Document
  datasources: DataSource[]
  binaryPaths: BinaryPaths
  testMode?: boolean
  copyRuntime?: boolean
  engineVersion: string
  clientVersion: string
  activeProvider: string
  dataProxy: boolean
}

export interface BuildClientResult {
  fileMap: Dictionary<string>
  prismaClientDmmf: PrismaClientDMMF.Document
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function buildClient({
  schemaPath,
  runtimeDirs,
  binaryPaths,
  outputDir,
  generator,
  dmmf,
  datasources,
  engineVersion,
  clientVersion,
  projectRoot,
  activeProvider,
  dataProxy,
}: O.Required<GenerateClientOptions, 'runtimeDirs'>): Promise<BuildClientResult> {
  // we define the basic options for the client generation
  const document = getPrismaClientDMMF(dmmf)
  const clientEngineType = getClientEngineType(generator!)
  const tsClientOptions = {
    document,
    datasources,
    generator,
    platforms:
      clientEngineType === ClientEngineType.Library
        ? (Object.keys(binaryPaths.libqueryEngine ?? {}) as Platform[])
        : (Object.keys(binaryPaths.queryEngine ?? {}) as Platform[]),
    schemaPath,
    outputDir,
    clientVersion,
    engineVersion,
    projectRoot: projectRoot!,
    activeProvider,
    dataProxy,
  }

  // we create a regular client that is fit for Node.js
  const nodeTsClient = new TSClient({
    ...tsClientOptions,
    runtimeName: 'index',
    runtimeDir: runtimeDirs.node,
  })

  // we create a client that is fit for edge runtimes
  const edgeTsClient = new TSClient({
    ...tsClientOptions,
    dataProxy: true, // edge only works w/ data proxy
    runtimeName: 'edge',
    runtimeDir: runtimeDirs.edge,
  })

  const fileMap = {} // we will store the generated contents here

  // we generate the default client that is meant to work on Node
  fileMap['index.js'] = await JS(nodeTsClient, false)
  fileMap['index.d.ts'] = await TS(nodeTsClient)
  fileMap['index-browser.js'] = await BrowserJS(nodeTsClient)
  fileMap['package.json'] = JSON.stringify(
    {
      name: GENERATED_PACKAGE_NAME,
      main: 'index.js',
      types: 'index.d.ts',
      browser: 'index-browser.js',
    },
    null,
    2,
  )

  // we only generate the edge client if `--data-proxy` is passed
  if (dataProxy === true) {
    fileMap['edge.js'] = await JS(edgeTsClient, true)
    fileMap['edge.d.ts'] = await TS(edgeTsClient, true)
  }

  if (generator?.previewFeatures.includes('deno') && !!globalThis.Deno) {
    if (dataProxy === true) {
      // we create a client that is fit for edge runtimes
      const denoEdgeTsClient = new TSClient({
        ...tsClientOptions,
        dataProxy: true, // edge only works w/ data proxy
        runtimeName: 'index.d.ts',
        runtimeDir: '../' + runtimeDirs.edge,
        deno: true,
      })

      fileMap['deno/edge.js'] = await JS(denoEdgeTsClient, true)
      fileMap['deno/index.d.ts'] = await TS(denoEdgeTsClient)
      fileMap['deno/edge.ts'] = `
import './polyfill.js'
// @deno-types="./index.d.ts"
export * from './edge.js'`
      fileMap['deno/polyfill.js'] = 'globalThis.process = { env: Deno.env.toObject() }; globalThis.global = globalThis'
    }
  }

  return {
    fileMap, // a map of file names to their contents
    prismaClientDmmf: document, // the DMMF document
  }
}

// TODO: explore why we have a special case for excluding pnpm
async function getDefaultOutdir(outputDir: string): Promise<string> {
  if (outputDir.endsWith('node_modules/@prisma/client')) {
    return path.join(outputDir, '../../.prisma/client')
  }
  if (
    process.env.INIT_CWD &&
    process.env.npm_lifecycle_event === 'postinstall' &&
    !process.env.PWD?.includes('.pnpm')
  ) {
    // INIT_CWD is the dir, in which "npm install" has been invoked. That can e.g. be in ./src
    // If we're in ./ - there'll also be a package.json, so we can directly go for it
    // otherwise, we'll go up in the filesystem and look for the first package.json
    if (fs.existsSync(path.join(process.env.INIT_CWD, 'package.json'))) {
      return path.join(process.env.INIT_CWD, 'node_modules/.prisma/client')
    }
    const packagePath = await pkgUp({ cwd: process.env.INIT_CWD })
    if (packagePath) {
      return path.join(path.dirname(packagePath), 'node_modules/.prisma/client')
    }
  }

  return path.join(outputDir, '../../.prisma/client')
}

export async function generateClient(options: GenerateClientOptions): Promise<void> {
  const {
    datamodel,
    schemaPath,
    outputDir,
    transpile,
    generator,
    dmmf,
    datasources,
    binaryPaths,
    testMode,
    copyRuntime,
    clientVersion,
    engineVersion,
    activeProvider,
    dataProxy,
  } = options

  const clientEngineType = getClientEngineType(generator!)
  const { runtimeDirs, finalOutputDir, projectRoot } = await getGenerationDirs(options)

  const { prismaClientDmmf, fileMap } = await buildClient({
    datamodel,
    schemaPath,
    transpile,
    runtimeDirs,
    outputDir: finalOutputDir,
    generator,
    dmmf,
    datasources,
    binaryPaths,
    clientVersion,
    engineVersion,
    projectRoot,
    activeProvider,
    dataProxy,
  })

  const denylistsErrors = validateDmmfAgainstDenylists(prismaClientDmmf)

  if (denylistsErrors) {
    let message = `${chalk.redBright.bold(
      'Error: ',
    )}The schema at "${schemaPath}" contains reserved keywords.\n       Rename the following items:`

    for (const error of denylistsErrors) {
      message += '\n         - ' + error.message
    }

    message += `\nTo learn more about how to rename models, check out https://pris.ly/d/naming-models`

    throw new DenylistError(message)
  }

  await ensureDir(finalOutputDir)
  await ensureDir(path.join(outputDir, 'runtime'))
  if (generator?.previewFeatures.includes('deno') && !!globalThis.Deno) {
    await ensureDir(path.join(outputDir, 'deno'))
  }
  // TODO: why do we sometimes use outputDir and sometimes finalOutputDir?
  // outputDir:       /home/millsp/Work/prisma/packages/client
  // finalOutputDir:  /home/millsp/Work/prisma/.prisma/client

  await Promise.all(
    Object.entries(fileMap).map(async ([fileName, file]) => {
      const filePath = path.join(finalOutputDir, fileName)
      // The deletion of the file is necessary, so VSCode
      // picks up the changes.
      if (await exists(filePath)) {
        await remove(filePath)
      }
      await writeFile(filePath, file)
    }),
  )
  const runtimeSourceDir = testMode
    ? eval(`require('path').join(__dirname, '../../runtime')`)
    : eval(`require('path').join(__dirname, '../runtime')`)

  // if users use a custom output dir
  if (copyRuntime || !path.resolve(outputDir).endsWith(`@prisma${path.sep}client`)) {
    // TODO: Windows, / is not working here...
    const copyTarget = path.join(outputDir, 'runtime')
    await ensureDir(copyTarget)
    if (runtimeSourceDir !== copyTarget) {
      await copy({
        from: runtimeSourceDir,
        to: copyTarget,
        recursive: true,
        parallelJobs: process.platform === 'win32' ? 1 : 20,
        overwrite: true,
      })
    }
  }
  const enginePath =
    clientEngineType === ClientEngineType.Library ? binaryPaths.libqueryEngine : binaryPaths.queryEngine

  if (!enginePath) {
    throw new Error(
      `Prisma Client needs \`${
        clientEngineType === ClientEngineType.Library ? 'libqueryEngine' : 'queryEngine'
      }\` in the \`binaryPaths\` object.`,
    )
  }

  if (transpile === true && dataProxy !== true) {
    if (process.env.NETLIFY) {
      await ensureDir('/tmp/prisma-engines')
    }

    for (const [binaryTarget, filePath] of Object.entries(enginePath)) {
      const fileName = path.basename(filePath)
      const target =
        process.env.NETLIFY && binaryTarget !== 'rhel-openssl-1.0.x'
          ? path.join('/tmp/prisma-engines', fileName)
          : path.join(finalOutputDir, fileName)
      const [sourceFileSize, targetFileSize] = await Promise.all([fileSize(filePath), fileSize(target)])

      // If the target doesn't exist yet, copy it
      if (!targetFileSize) {
        if (fs.existsSync(filePath)) {
          await overwriteFile(filePath, target)
          continue
        } else {
          throw new Error(`File at ${filePath} is required but was not present`)
        }
      }

      // If target !== source size, they're definitely different, copy it
      if (targetFileSize && sourceFileSize && targetFileSize !== sourceFileSize) {
        await overwriteFile(filePath, target)
        continue
      }
      const binaryName =
        clientEngineType === ClientEngineType.Binary ? BinaryType.queryEngine : BinaryType.libqueryEngine
      // They must have an equal size now, let's check for the hash
      const [sourceVersion, targetVersion] = await Promise.all([
        getEngineVersion(filePath, binaryName).catch(() => null),
        getEngineVersion(target, binaryName).catch(() => null),
      ])

      if (sourceVersion && targetVersion && sourceVersion === targetVersion) {
        // skip
      } else {
        await overwriteFile(filePath, target)
      }
    }
  }

  const schemaTargetPath = path.join(finalOutputDir, 'schema.prisma')
  if (schemaPath !== schemaTargetPath) {
    await copyFile(schemaPath, schemaTargetPath)
  }

  const proxyIndexJsPath = path.join(outputDir, 'index.js')
  const proxyIndexBrowserJsPath = path.join(outputDir, 'index-browser.js')
  const proxyIndexDTSPath = path.join(outputDir, 'index.d.ts')
  if (!fs.existsSync(proxyIndexJsPath)) {
    await copyFile(path.join(__dirname, '../../index.js'), proxyIndexJsPath)
  }

  if (!fs.existsSync(proxyIndexDTSPath)) {
    await copyFile(path.join(__dirname, '../../index.d.ts'), proxyIndexDTSPath)
  }

  if (!fs.existsSync(proxyIndexBrowserJsPath)) {
    await copyFile(path.join(__dirname, '../../index-browser.js'), proxyIndexBrowserJsPath)
  }
}

async function fileSize(name: string): Promise<number | null> {
  try {
    const statResult = await stat(name)
    return statResult.size
  } catch (e) {
    return null
  }
}

function validateDmmfAgainstDenylists(prismaClientDmmf: PrismaClientDMMF.Document): Error[] | null {
  const errorArray = [] as Error[]

  const denylists = {
    // A copy of this list is also in prisma-engines. Any edit should be done in both places.
    // https://github.com/prisma/prisma-engines/blob/main/psl/parser-database/src/names/reserved_model_names.rs
    models: [
      // Reserved Prisma keywords
      'PrismaClient',
      'Prisma',
      // JavaScript keywords
      'break',
      'case',
      'catch',
      'class',
      'const',
      'continue',
      'debugger',
      'default',
      'delete',
      'do',
      'else',
      'enum',
      'export',
      'extends',
      'false',
      'finally',
      'for',
      'function',
      'if',
      'implements',
      'import',
      'in',
      'instanceof',
      'interface',
      'let',
      'new',
      'null',
      'package',
      'private',
      'protected',
      'public',
      'return',
      'super',
      'switch',
      'this',
      'throw',
      'true',
      'try',
      'typeof',
      'var',
      'void',
      'while',
      'with',
      'yield',
    ],
    fields: ['AND', 'OR', 'NOT'],
    dynamic: [],
  }

  if (prismaClientDmmf.datamodel.enums) {
    for (const it of prismaClientDmmf.datamodel.enums) {
      if (denylists.models.includes(it.name) || denylists.fields.includes(it.name)) {
        errorArray.push(Error(`"enum ${it.name}"`))
      }
    }
  }

  if (prismaClientDmmf.datamodel.models) {
    for (const it of prismaClientDmmf.datamodel.models) {
      if (denylists.models.includes(it.name) || denylists.fields.includes(it.name)) {
        errorArray.push(Error(`"model ${it.name}"`))
      }
    }
  }

  return errorArray.length > 0 ? errorArray : null
}

/**
 * Get all the directories involved in the generation process.
 *
 * @param useDefaultOutdir if we are generating to the default output
 * @param runtimeDirs overrides for the runtime directories
 * @returns
 */
async function getGenerationDirs({
  testMode,
  runtimeDirs,
  generator,
  outputDir,
  datamodel,
  schemaPath,
}: GenerateClientOptions) {
  const useDefaultOutdir = testMode ? !runtimeDirs : !generator?.isCustomOutput

  const _runtimeDirs = {
    // if we have an override, we use it, but if not then use the defaults
    node: runtimeDirs?.node || (useDefaultOutdir ? '@prisma/client/runtime' : './runtime'),
    edge: runtimeDirs?.edge || (useDefaultOutdir ? '@prisma/client/runtime' : './runtime'),
  }

  const finalOutputDir = useDefaultOutdir ? await getDefaultOutdir(outputDir) : outputDir
  if (!useDefaultOutdir) {
    await verifyOutputDirectory(finalOutputDir, datamodel, schemaPath)
  }

  const packageRoot = await pkgUp({ cwd: path.dirname(finalOutputDir) })
  const projectRoot = packageRoot ? path.dirname(packageRoot) : process.cwd()

  return {
    runtimeDirs: _runtimeDirs,
    finalOutputDir,
    projectRoot,
  }
}

async function verifyOutputDirectory(directory: string, datamodel: string, schemaPath: string) {
  let content: string
  try {
    content = await fs.promises.readFile(path.join(directory, 'package.json'), 'utf8')
  } catch (e) {
    if (e.code === 'ENOENT') {
      // no package.json exists, we are good
      return
    }
    throw e
  }
  const { name } = JSON.parse(content)
  if (name === clientPackageName) {
    const message = [`Generating client into ${chalk.bold(directory)} is not allowed.`]
    message.push('This package is used by `prisma generate` and overwriting its content is dangerous.')
    message.push('')
    message.push('Suggestion:')
    const outputDeclaration = findOutputPathDeclaration(datamodel)

    if (outputDeclaration && outputDeclaration.content.includes(clientPackageName)) {
      const outputLine = outputDeclaration.content
      message.push(`In ${chalk.bold(schemaPath)} replace:`)
      message.push('')
      message.push(
        `${chalk.dim(outputDeclaration.lineNumber)} ${replacePackageName(outputLine, chalk.red(clientPackageName))}`,
      )
      message.push('with')

      message.push(
        `${chalk.dim(outputDeclaration.lineNumber)} ${replacePackageName(outputLine, chalk.green('.prisma/client'))}`,
      )
    } else {
      message.push(
        `Generate client into ${chalk.bold(replacePackageName(directory, chalk.green('.prisma/client')))} instead`,
      )
    }

    message.push('')
    message.push("You won't need to change your imports.")
    message.push('Imports from `@prisma/client` will be automatically forwarded to `.prisma/client`')
    const error = new Error(message.join('\n'))
    throw error
  }
}

function replacePackageName(directoryPath: string, replacement: string): string {
  return directoryPath.replace(clientPackageName, replacement)
}

function findOutputPathDeclaration(datamodel: string): OutputDeclaration | null {
  const lines = datamodel.split(/\r?\n/)

  for (const [i, line] of lines.entries()) {
    if (/output\s*=/.test(line)) {
      return { lineNumber: i + 1, content: line.trim() }
    }
  }
  return null
}
