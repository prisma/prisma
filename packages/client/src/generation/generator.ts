import Debug from '@prisma/debug'
import { enginesVersion } from '@prisma/engines-version'
import { generatorHandler } from '@prisma/generator-helper'
import { ClientEngineType, getClientEngineType, parseEnvValue } from '@prisma/internals'

import { externalToInternalDmmf } from '../runtime/externalToInternalDmmf'
import { generateClient } from './generateClient'
import { getDMMF } from './getDMMF'
import { dmmfToTypes } from './utils/types/dmmfToTypes'

const debug = Debug('prisma:client:generator')

// See https://www.notion.so/prismaio/Prisma-Generators-a2cdf262207a4e9dbcd0e362dfac8dc0

const pkg = require('../../package.json')

const clientVersion = pkg.version

// if the file has been run as a CLI
if (process.argv[1] === __filename) {
  generatorHandler({
    onManifest(config) {
      const requiredEngine = getClientEngineType(config) === ClientEngineType.Library ? 'libqueryEngine' : 'queryEngine'
      debug(`requiredEngine: ${requiredEngine}`)
      return {
        defaultOutput: '.prisma/client', // the value here doesn't matter, as it's resolved in https://github.com/prisma/prisma/blob/88fe98a09092d8e53e51f11b730c7672c19d1bd4/packages/sdk/src/get-generators/getGenerators.ts
        prettyName: 'Prisma Client',
        requiresEngines: [requiredEngine],
        version: clientVersion,
        requiresEngineVersion: enginesVersion,
      }
    },
    async onGenerate(options) {
      // CLI versions < 2.20.0 still send a string
      // CLIs >= 2.20.0 send an `EnvValue`
      const outputDir =
        typeof options.generator.output === 'string'
          ? options.generator.output
          : parseEnvValue(options.generator.output!)

      return generateClient({
        datamodel: options.datamodel,
        schemaPath: options.schemaPath,
        binaryPaths: options.binaryPaths!,
        datasources: options.datasources,
        outputDir,
        copyRuntime: Boolean(options.generator.config.copyRuntime),
        dmmf: options.dmmf,
        generator: options.generator,
        engineVersion: options.version,
        clientVersion,
        transpile: true,
        activeProvider: options.datasources[0]?.activeProvider,
        dataProxy: options.dataProxy,
      })
    },
  })
}

export { dmmfToTypes, externalToInternalDmmf, getDMMF }
