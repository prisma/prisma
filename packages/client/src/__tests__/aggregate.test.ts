import { getDMMF } from '../generation/getDMMF'
import { DMMFClass, makeDocument } from '../runtime'

export const recommender = /* Prisma */ `
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model Article {
  id      Int      @id
  url     String   @unique
  title   String
  clicks  Int
  content String
  date    DateTime
  likedBy User[]
  link    Link?
}

model Link {
  id        Int      @id
  articleId Int      @unique
  article   Article  @relation(fields: [articleId], references: [id])
  postedAt  DateTime
}

model User {
  id            Int       @id
  name          String
  email         String    @unique
  likedArticles Article[]
  age           Int?
  personaId     Int
  persona       Persona   @relation(fields: [personaId], references: [id])
}

model Persona {
  id          Int     @id
  isDeveloper Boolean
  User        User[]
}
`

let dmmf
describe('aggregate', () => {
  beforeAll(async () => {
    dmmf = new DMMFClass(
      await getDMMF({
        datamodel: recommender,
      }),
    )
  })

  test('count happy path', () => {
    const document = makeDocument({
      dmmf,
      select: {
        take: 10,
        select: {
          _count: {
            select: {
              _all: true,
            },
          },
        },
      },
      rootTypeName: 'query',
      rootField: 'aggregateUser',
    })
    document.validate(undefined, false, 'user', 'colorless')
    expect(String(document)).toMatchInlineSnapshot(`
      query {
        aggregateUser(take: 10) {
          _count {
            _all
          }
        }
      }
    `)
  })

  test('combined happy path', () => {
    const document = makeDocument({
      dmmf,
      select: {
        take: 10,
        cursor: {
          email: 'a@a.de',
        },
        orderBy: {
          age: 'asc',
        },
        skip: 12,
        where: {
          age: { gt: 500 },
        },
        select: {
          _count: true,
          _avg: {
            select: {
              age: true,
            },
          },
          _min: {
            select: {
              age: true,
            },
          },
          _max: {
            select: {
              age: true,
            },
          },
          _sum: {
            select: {
              age: true,
            },
          },
        },
      },
      rootTypeName: 'query',
      rootField: 'aggregateUser',
    })
    document.validate(undefined, false, 'user', 'colorless')
    expect(String(document)).toMatchSnapshot()
  })

  test('unhappy path - incorrect arg', () => {
    const select = { mount: true }
    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'query',
      rootField: 'aggregateUser',
    })
    expect(() => document.validate(select, false, 'user', 'colorless')).toThrowErrorMatchingSnapshot()
  })

  test('unhappy path - incorrect field', () => {
    const select = {
      select: {
        _avg: {
          select: {
            blub: true,
          },
        },
      },
    }
    const document = makeDocument({
      dmmf,
      select,
      rootTypeName: 'query',
      rootField: 'aggregateUser',
    })
    expect(() => document.validate(select, false, 'user', 'colorless')).toThrowErrorMatchingSnapshot()
  })
})
