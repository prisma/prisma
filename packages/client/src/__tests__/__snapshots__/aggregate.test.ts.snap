// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aggregate combined happy path 1`] = `
query {
  aggregateUser(
    take: 10
    cursor: {
      email: "a@a.de"
    }
    orderBy: [
      {
        age: asc
      }
    ]
    skip: 12
    where: {
      age: {
        gt: 500
      }
    }
  ) {
    _count {
      id
      name
      email
      age
      personaId
      _all
    }
    _avg {
      age
    }
    _min {
      age
    }
    _max {
      age
    }
    _sum {
      age
    }
  }
}
`;

exports[`aggregate unhappy path - incorrect arg 1`] = `

Invalid \`prisma.user()\` invocation:

{
  mount: true
  ~~~~~
}

Unknown arg \`mount\` in mount for type AggregateUser. Did you mean \`where\`? Available args:
type aggregateUser {
  where?: UserWhereInput
  orderBy?: List<UserOrderByWithRelationInput> | UserOrderByWithRelationInput
  cursor?: UserWhereUniqueInput
  take?: Int
  skip?: Int
}


`;

exports[`aggregate unhappy path - incorrect field 1`] = `

Invalid \`prisma.user()\` invocation:

{
  select: {
    _avg: {
      select: {
        blub: true,
        ~~~~
?       id?: true,
?       age?: true,
?       personaId?: true
      }
    }
  }
}


Unknown field \`blub\` for select statement on model UserAvgAggregateOutputType. Available options are listed in green. Did you mean \`id\`?

`;
